Support TA profiling with gprof

jforissier opened the pull request:
Consolidating my RFC PR: https://github.com/OP-TEE/optee_os/pull/1128.
Depends on https://github.com/OP-TEE/optee_client/pull/61.

jenswi-linaro commented on core/include/kernel/tee_ta_manager.h:
```diff
@@ -152,4 +170,12 @@ void tee_ta_dump_current(void);
 
+#if defined(CFG_TA_GPROF_SUPPORT)
+void tee_ta_gprof_sample_pc(vaddr_t pc);
+void tee_ta_update_session_utime(bool suspend);
```
The code would be easier to follow if this function either took a descriptive enum instead of a bool or if there was two functions, `tee_ta_update_session_utime_suspend()` and `tee_ta_update_session_utime_resume()`. I'm leaning towards the latter, but it's no big deal.

jforissier replied:
OK will add suspend/resume.

jenswi-linaro commented on core/arch/arm/sta/gprof.c:
```diff
@@ -0,0 +1,324 @@
+static TEE_Result gprof_start_pc_sampling(uint32_t param_types,
+					  TEE_Param params[TEE_NUM_PARAMS])
+{
+
```
Empty line, not needed.

jforissier replied:
OK

jenswi-linaro commented on core/arch/arm/sta/gprof.c:
```diff
@@ -0,0 +1,324 @@
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+
+	sbuf->nsamples = len/sizeof(*sbuf->samples);
```
Please add space around `/`

jenswi-linaro commented on core/arch/arm/sta/gprof.c:
```diff
@@ -0,0 +1,324 @@
+	cache_maintenance_l1(DCACHE_AREA_CLEAN, (void *)va, size);
+	cache_maintenance_l1(ICACHE_AREA_INVALIDATE, (void *)va, size);
+
+	params[1].value.a = va & 0xFFFFFFFF;
```
Please use `reg_pair_from_64()` from `<kernel/misc.h>` instead.

jforissier replied:
OK

jenswi-linaro commented on core/arch/arm/sta/gprof.c:
```diff
@@ -0,0 +1,324 @@
+	pa = tee_mm_get_smem(mm);
+	va = tee_mmu_map_gprof_heap(utc, pa, size);
+
+	tee_mmu_set_ctx(&utc->ctx);
```
This will shift the virtual addresses of mapped parameters in the current invoke of the TA. The pointers to buffer in the TA arguments will suddenly point into the gprof heap instead. I think we need a better way of doing this.

jforissier replied:
Yeah :( Should we put the heap at some fixed VA maybe? For instance, at the top of the 4G VA space?

jenswi-linaro replied:
That will cost another page table. Can it be dynamically reassigned between invokes?

jforissier replied:
Not really I guess, because libutee has a pointer to that buffer and it's used throughout the life of the TA (from the point profiling is started, until the last session is closed).
It's in `static struct gmon_param` [here](https://github.com/OP-TEE/optee_os/pull/1209/commits/c6b326e7358b8e6d11351ab69c57b4cb03745c6a#diff-6d699c6cedecfb7007bbd8b2df37d73eR76), set by `__gprof_ini()` [here](https://github.com/OP-TEE/optee_os/pull/1209/commits/c6b326e7358b8e6d11351ab69c57b4cb03745c6a#diff-6d699c6cedecfb7007bbd8b2df37d73eR144).

jenswi-linaro replied:
As I understand it this is normally initialized during open session. If we place it right after this parameter mapping and then in the following invokes places the parameter mapping after this we'll use the va-space reasonably efficiently. There may be a va-hole just before the gprof heap, but as you normally don't supply that much data to an open session it's expected to be small.

jforissier replied:
No, not always during open session. It is initialized on-demand, when the first instrumented function is called (or never, if the TA is not instrumented). That might be much later than session open (if only some source file of the TA is instrumented), or even sooner (if libutils/libutee is instrumented). We could try to force the init to be done at the time the session is opened, but it would mean being able to detect that profiling was enabled in the binary. Not sure how to do that.

What's the problem with reserving a page table and the upper VA space for the heap? In fact, it could be a first step towards a truly dynamic heap for TAs, I mean getting rid of `ta_heap` in `user_ta_header.c` and relying on a TEE core service to provide VA space instead (similar to the linux libc calling `sbrk()` actually).

jenswi-linaro replied:
The problem is spelled "4 kB of secure kernel memory" for LPAE. What we need is better management of TA va-space. Using a fixed address at the end of the TA va-space wastes as much secure kernel memory as possible.

I don't think we should have a dynamic heap as we still need to be able to guarantee that the specified amount of heap is available.

jforissier replied:
> The problem is spelled "4 kB of secure kernel memory" for LPAE. What we need is better management of TA va-space. Using a fixed address at the end of the TA va-space wastes as much secure kernel memory as possible.

OK, I see.

> I don't think we should have a dynamic heap as we still need to be able to guarantee that the specified amount of heap is available.

Aside from the overhead you mentioned above, I don't see what difference it would make. Suppose `ta_header_add_session()` would call a kernel function to obtain VA space for the heap just before doing `malloc_add_pool()`, something like:
```diff
 static TEE_Result ta_header_add_session(uint32_t session_id)
 {
         struct ta_session *itr = ta_header_get_session(session_id);

         if (itr)
                 return TEE_SUCCESS;

         ta_ref_count++;

         if (ta_ref_count == 1) {
                 TEE_Result res;

                 if (!context_init) {
                         trace_set_level(tahead_get_trace_level());
+                        ta_heap = get_va_space_for_heap(ta_heap_size);
+                        if (!ta_heap)
+                                 return TEE_ERROR_OUT_OF_MEMORY;
                         malloc_add_pool(ta_heap, ta_heap_size);
                         _TEE_MathAPI_Init();
                         context_init = true;
                 }
```
WIth this code, if the creation of the first session succeeds, then all the subsequent ones will be guaranteed to have the required heap. AFAICT it does not change the current behavior.

I need to think more, how we can detect that the TA is instrumented at load time, so that the allocation can be done early. In which case we could put the gprof heap between the TA code and the parameter mappings I suppose.

jenswi-linaro replied:
Yes, allocating heap memory like that is OK.

jenswi-linaro commented on core/arch/arm/sta/gprof.c:
```diff
@@ -0,0 +1,324 @@
+	va = tee_mmu_map_gprof_heap(utc, pa, size);
+
+	tee_mmu_set_ctx(&utc->ctx);
+	cache_maintenance_l1(DCACHE_AREA_CLEAN, (void *)va, size);
```
No need for this, we don't have any runnable code in the gprof heap.

jforissier replied:
You mean no need for `ICACHE_AREA_INVALIDATE`, right? But, should I also remove `DCACHE_AREA_CLEAN`?

jenswi-linaro replied:
No need to clean the d-cache either.

jforissier replied:
OK

jenswi-linaro commented on ta/arch/arm/ta.ld.S:
```diff
@@ -32,12 +32,32 @@ SECTIONS {
 		/* Workaround for an erratum in ARM's VFP11 coprocessor */
 		*(.vfp11_veneer)
+#ifdef ARM32
+		PROVIDE(__gnu_mcount_nc = __gnu_mcount_nc__utee);
```
Wouldn't it work to use a define instead of `__gnu_mcount_nc`/`_mcount` to get rid of the ifdeffing here and below?

jforissier replied:
Yes, I think it should work.

jenswi-linaro commented on lib/libutee/arch/arm/gprof/gprof_a64.S:
```diff
@@ -38,21 +38,23 @@
+ * script if the TA is instrumented, i.e., if _mcount is referenced
  */
-FUNC _mcount, :
+FUNC _mcount__utee, :
```
We should have double leading `_` for proper namespacing.

jforissier replied:
OK

jenswi-linaro commented on lib/libutee/arch/arm/gprof/gprof_a32.S:
```diff
@@ -39,19 +39,20 @@
+ * linker script, only if__gnu_mcount_nc is referenced
  */
-FUNC __gnu_mcount_nc, :
+FUNC __gnu_mcount_nc__utee, :
```
Can we have the same name as in the AArch64 case here?

jenswi-linaro commented on lib/libutee/arch/arm/gprof/gprof.c:
```diff
@@ -73,6 +73,31 @@
 #include "gmon_out.h"
 #include "gprof_pta.h"
 
+extern uint8_t _mcount_referenced[]; /* Exported by the TA linker script */
```
Please use double leading `_` for proper namespacing.

jforissier commented:
Doh, you're right, we're almost back to square one :( Or maybe not. In fact I still think it is a good idea to initialize gprof as early as possible. Like the normal world libc does, actually.

What about adding an explicit initialization call into libutee? I'm thinking about:
```diff
diff --git a/lib/libutee/include/utee_types.h b/lib/libutee/include/utee_types.h
index 307b507..9d1fcc5 100644
--- a/lib/libutee/include/utee_types.h
+++ b/lib/libutee/include/utee_types.h
@@ -41,6 +41,8 @@ enum utee_entry_func {
        UTEE_ENTRY_FUNC_OPEN_SESSION = 0,
        UTEE_ENTRY_FUNC_CLOSE_SESSION,
        UTEE_ENTRY_FUNC_INVOKE_COMMAND,
+       UTEE_ENTRY_FUNC_INIT,
 };
```

The TA entry point would be called by TEE Core once with this new value, without any additional parameter, and before any other call is done.
Do you see any other option? Such as call `__utee_gprof_init()` earlier, in `entry_invoke_command()`, and if the mapping changed, fixup the `struct utee_params *up` parameter and account for the shift? That would be quite ugly I guess.
Any other alternative?

jenswi-linaro commented:
I was also thinking about a `UTEE_ENTRY_FUNC_INIT`, please add one.

jforissier commented:
OK I can try that way. But, thinking a bit more about all this, I am tempted to propose a radical simplification...

Going back to the initial issue: in `__utee_gprof_init()` we need a user-mode buffer, which would typically be allocated via malloc(). The problem is, the TA heap size is controlled by the application developer (via ta_heap_size, or actually the TA_DATA_SIZE define), and it is therefore likely to be too small when profiling is enabled. And we have no mean to ask the TEE core to extend the TA heap.

Now, what we could do is just add a compile-time flag, TA_DATA_SIZE_MULT, default value 1, and in `ta/arch/arm/user_ta_header.c`:
```C
uint8_t ta_heap[TA_DATA_SIZE * TA_DATA_SIZE_MULT];
```
Then allocate the gprof call graph buffer with plain `malloc()` and scratch the whole special kernel alloc and mapping. When compiling an instrumented TA, in addition to `-pg`, just set TA_DATA_SIZE_MULT to 2 or 3 or whatever it takes. Problem solved. If later we implement a more clever malloc heap that can `mmap()` more memory, the code won't change, the multiplier may become useless that's it. Another advantage is that it's not kernel-alloc'ed memory that is mapped.

Thoughts?

jenswi-linaro commented:
Then you could use a uint8_t ta_pgheap[TA_PGHEAP_SIZE] instead. Perhaps with a warning in the link script if the size it far too small.

jforissier commented:
Updated. By reserving space in the linker script, I don't even need to change `user_ta_header.c`.

jenswi-linaro commented on mk/config.mk:
```diff
@@ -199,3 +199,24 @@ CFG_BOOT_SECONDARY_REQUEST ?= n
+
+ifeq ($(CFG_TA_GPROF),y)
+# Build various user-mode libraries with profiling enabled (-pg)
+CFG_LIBUTEE_GPROF ?= y
```
This seems complex, why not build all user TA libs with profiling support if `CFG_LIBUTEE_GPROF` is `y`?
If we discover that we need more flexibility, let's deal with that then.

jforissier replied:
Sounds reasonable. I'll use `CFG_ULIBS_GPROF` instead (because `LIBUTEE` sounds too restrictive). Feel free to propose a better name ;)

jenswi-linaro commented on core/arch/arm/sta/gprof.c:
```diff
@@ -0,0 +1,275 @@
+	if (!sbuf)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	sbuf->samples = calloc(1, len);
```
Wouldn't it be nice to keep this buffer in user space too? Especially since user space decides the size.

jforissier replied:
Could the TEE core still access the buffer from `thread_state_suspend()` -> `tee_ta_gprof_sample_pc(pc)`? We'd have to make sure it's always mapped, no?

jenswi-linaro replied:
Yes, I'm quite sure it's mapped at that point. The VA space associated with a session/context should always be available except while switching session/context.

jforissier replied:
Yep that works.

jforissier commented on core/arch/arm/include/kernel/user_ta.h:
```diff
@@ -53,6 +53,9 @@ struct user_ta_ctx {
 	struct tee_storage_enum_head storage_enums;
 	tee_mm_entry_t *mm;	/* secure world memory */
 	tee_mm_entry_t *mm_stack;/* stack */
+#ifdef CFG_TA_GPROF_SUPPORT
```
Yes, will fix

jforissier commented on core/arch/arm/sta/gprof.c:
```diff
@@ -0,0 +1,275 @@
+{
+	TEE_Result res;
+
+	res = tee_ta_get_current_session(s);
```
Leftover from code that was structured differently. I'll simply remove this function and use `tee_ta_get_calling_session()` directly.

jforissier commented on core/arch/arm/sta/gprof.c:
```diff
@@ -0,0 +1,275 @@
+	if (exp_pt != param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = get_calling_session(&s);
```
OK

jforissier replied:
My understanding is that `tee_ta_get_calling_session()` returns NULL if we're called from a client app, or non-NULL if we're called from a TA (in which case, `s->clnt_id.login` is necessarily `TEE_LOGIN_TRUSTED_APP` but I can check that also). So the "nonsecure client" check is already handled.

Now for the "static TA" check, I think I can add `!is_static_ta_ctx(s->ctx);`. I'll try that.

jforissier commented on core/include/kernel/tee_ta_manager.h:
```diff
@@ -65,6 +65,21 @@ struct tee_ta_ops {
+	uint64_t usr_entered;	/* When this session last entered user mode */
+	uint32_t freq;		/* @usr divided by @freq is in seconds */
+};
+
```
Yup

jforissier commented on core/include/kernel/tee_ta_manager.h:
```diff
@@ -152,4 +170,14 @@ void tee_ta_dump_current(void);
+void tee_ta_update_session_utime_resume(void);
+#else
+static inline void tee_ta_gprof_sample_pc(vaddr_t pc __unused) {}
+static inline void tee_ta_update_session_utime_suspend(void);
```
Nothing, it's just silly and won't compile ;) I'll fix that.

jenswi-linaro replied:
OK, I thought I was about to learn something... :-)

jforissier commented on core/kernel/tee_ta_manager.c:
```diff
@@ -713,3 +717,63 @@ void tee_ta_dump_current(void)
+	struct sample_buf *sbuf;
+	size_t idx;
+
+	tee_ta_get_current_session(&s);
```
OK

jforissier commented on core/kernel/tee_ta_manager.c:
```diff
@@ -713,3 +717,63 @@ void tee_ta_dump_current(void)
+	struct sample_buf *sbuf;
+	uint64_t now;
+
+	tee_ta_get_current_session(&s);
```
OK

jforissier commented:
Updated. Now the whole profiling buffer is reserved in `.bss` by the linker script. Tested on QEMU and D02 with 64- and 32-bit TAs.

jenswi-linaro commented on core/kernel/tee_ta_manager.c:
```diff
@@ -344,6 +344,11 @@ TEE_Result tee_ta_close_session(struct tee_ta_session *csess,
 	tee_ta_unlink_session(sess, open_sessions);
+#if defined(CFG_TA_GPROF_SUPPORT)
+       if (sess->sbuf)
+               free(sess->sbuf->samples);
```
Isn't `sess->sbuf->samples` a pointer to user space?

jforissier replied:
Good catch, I'll scratch that. I'm wondering why I got no error or crash however.

jenswi-linaro commented:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jforissier commented:
Rebased, fixed 4 checkpatch warnings, applied Jen's R-b, and added two commits for documentation and Travis.

jenswi-linaro commented:
Anything we're waiting for here?

jbech-linaro commented:
> @jenswi-linaro 
> Anything we're waiting for here?

The license question, whether this is compatible with BSD 2-Clause or not.

jforissier commented:
Rebased and fixed documentation. Can someone review the doc please?

jenswi-linaro commented on documentation/gprof.md:
```diff
@@ -0,0 +1,58 @@
+## Usage
+
+- Build OP-TEE OS with `CFG_TA_GPROF_SUPPORT=y`. You will likely need to
+  increase the TEE heap: for instance `CFG_CORE_HEAP_SIZE=262144`. You
```
Is this really true? Why does the core heap need to be increased so much?

jforissier replied:
I've used this value quite some time ago when testing on D02, but the memory allocation was a bit different than the final version.
I'll double check this and give more precise information (i.e., the additional amount of memory that is required as a function of the TA `.text` size).

jforissier commented:
Doc updated.

jenswi-linaro commented:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jforissier commented:
Review tags added. I have simplified the commit text about the performance impact (details were mostly irrelevant) and mentioned the memory overhead. And I have also corrected the memory overhead figures in the documentation: `.bss` is increased by 136% (or 177%) of the size of `.text`, not just by 36% (or 77%). Sorry for the multiple pushes.

Should be ready for merge. @jbech-linaro what do you say?

jbech-linaro commented:
> @jforissier 
> Should be ready for merge. @jbech-linaro what do you say?

Yes! Based on the reply from Dave, we're good to go.

jbech-linaro commented on documentation/gprof.md:
```diff
@@ -0,0 +1,60 @@
+other one is the callee) as well as the number of times this specific
+arc of the call graph has been invoked.
+
+2. PC distribution over time
```
In the rendered version this also becomes "1" and not two.

jbech-linaro replied:
In the rendered version this also becomes "1" and not two.

jforissier commented:
Documentation issue fixed.
